Update data analysis to use new dataset (#17)
duke-malaria-collaboratory/DiscovR: _episodes_rmd/03-r-data-analysis.Rmd

> ```
> Bonus 1: 
> ```{r}
> ggplot(smoking, aes(x=year,y=pop)) +
> geom_point() +
> facet_wrap(vars(continent)) +
> scale_y_log10()
> ```
> Bonus 2 example:
> ```{r}
> ggplot(smoking, aes(x=year,y=pop)) +
> geom_point() +
> facet_wrap(vars(continent)) +
> scale_y_log10() +
> labs(x = 'Year', y = 'Population') +
> theme_bw()
> ```
{: .solution}

It seems like there are a couple outliers. By the end of today, we'll be able to identify what countries those are. 

# Overview of the lesson
_[Back to top](#contents)_

Yesterday we spent a lot of time making plots in R using the ggplot2 package. Visualizing data using plots is a very powerful skill in R, but what if we would like to work with only a subset of our data? Or clean up messy data, calculate summary statistics, create a new variable, or join two datasets together? There are several different methods for doing this in R, and we will touch on a few today using functions the `dplyr` package, which is also part of the `tidyverse`.

# An introduction to data analysis in R using `dplyr` {#intro-data-analysis}

## Get stats fast with `summarize()` {#get-stats-fast-with-summarize}
[*Back to top*](#contents)

> Let's say we would like to know what is the mean (average) smoking rate in the dataset. R has a built in function function called `mean()` that will calculate this value for us. We can apply that function to our smoking_pct column using the `summarize()` function. Here's what that looks like:

Reviewer: ```suggestion
Let's say we would like to know the mean (average) smoking rate in the dataset. R has a built in function function called `mean()` that will calculate this value for us. We can apply that function to our smoking_pct column using the `summarize()` function. Here's what that looks like:
```

```{r AvgLifeExp}
summarize(smoking, mean_smoke_pct=mean(smoke_pct))
```

When we call `summarize()`, we can use any of the column names of our data object as values to pass to other functions. `summarize()` will return a new data object and our value will be returned as a column.

> **Note:** The `summarize()` and `summarise()` functions perform identical functions.

> We name this new column so we can use in a future argument. So the `mean_smoke_pct=` part tells `summarize()` to use "mean_smoke_pct" as the name of the new column. Note that you don't have to have quotes around this new name as long as it starts with a letter and doesn't include a space.

Reviewer: I see where you're going with this, but in this specific example, the mean_smoke_pct couldn't be used in a future argument?

Author: Not entirely sure what you mean, is this better?

```suggestion
The `mean_smoke_pct=` part tells `summarize()` to use "mean_smoke_pct" as the name of the new column. Note that you don't have to have quotes around this new name as long as it starts with a letter and doesn't include a space.
```

Reviewer: Yep!

> Instead of including the data as an argument, we can use the *pipe operator* `%>%` to pass the data value into the `summarize` function. 

Reviewer: ```suggestion
Instead of including the data object as an argument, we can use the *pipe operator* `%>%` to pass the data value into the `summarize` function. You can think of `%>%` as another way to type "and then."
```

```{r AvgLifeExpWithPipe}
smoking %>% summarize(mean_smoke_pct=mean(smoke_pct))
```
 
This line of code will do the exact same thing as our first summary command, but the piping function tells R to use the `smoking` dataframe as the first argument in the next function.

> This lets us "chain" together multiple functions, which will be helpful later. Note that the pipe (`%>%`) is a bit different from using the ggplot plus (`+`). Pipes take the output from the left side and use it as input to the right side. Plusses layer on additional information (right side) to a preexisting plot (left side). 

Reviewer: ```suggestion
This lets us "chain" together multiple functions, which will be helpful later. Note that the pipe (`%>%`) is a bit different from using the ggplot plus (`+`). Pipes take the output from the left side and use it as input to the right side. In other words, it tells R to do the function on the left *and then* the function on the right. In contrast, plusses layer on additional information (right side) to a preexisting plot (left side). 
```

> We can also add an <kdb>Enter</kdb> to make it look nicer:  

Reviewer: ```suggestion
We can also add an <kbd>Enter</kbd> to make it look nicer:  
```

```{r AvgLifeExpWithPipe2}
smoking %>%
  summarize(mean_smoke_pct=mean(smoke_pct))
```

> Using the *pipe operator* `%>%` and enter command makes our code more readable. The  *pipe operator* `%>%` also helps to avoid using nested function and minimizes the need for new variables.

Reviewer: ```suggestion
Using the *pipe operator* `%>%` and <kbd>Enter</kbd> makes our code more readable. The  *pipe operator* `%>%` also helps to avoid using nested functions and minimizes the need for new variables.
```

> Since we use the pipe operator so often, there is a keyboard shortcut for it in RStudio. You can press <kdb>Ctrl</kdb>+<kdb>Shift</kdb>+<kdb>M<kdb> on Windows or <kdb>Cmd<kdb>+<kdb>Shift<kdb>+<kdb>M<kdb> on a Mac.

Reviewer: No idea whether you need proper html here, but I'm adding it just in case.
```suggestion
Since we use the pipe operator so often, there is a keyboard shortcut for it in RStudio. You can press <kbd>Ctrl</kbd>+<kbd>Shift</kbd>+<kbd>M</kbd> on Windows or <kbd>Cmd</kbd>+<kbd>Shift</kbd>+<kbd>M</kbd> on a Mac.
```

To do this, we have to use `==` (double equals), which we use when testing if two values are equal to each other:

```{r, filtSummarize}
smoking %>%
  filter(year == 1990) %>%
  summarize(mean_smoke_pct_1990=mean(smoke_pct))
```

Notice how the pipe operator (`%>%`) allows us to combine these two simple steps into a more complicated data extraction? We took the data, filtered out the rows, then took the mean value. The argument we pass to `filter()` needs to be some expression that will return TRUE or FALSE. We can use comparisons like `>` (greater than) and `<` (less than) for example. Here we tested for equality using a double equals sign `==`. 

> ## Filtering the dataset
>
> What is the average percent of people with lung cancer for the most recent year in the dataset? *Hint: the column headers identified by `read_csv()` showed us there was a column called `lung_cancer_pct` in the dataset*
>
> Bonus: Figure out a way to write the code so you don't have to type the year into filter. 
> (Typing in the year is an example of hard-coding and we try to avoid it because if the data changes the number might also change.)
>
> > ## Solution
> >

> > > Identify the earliest year in our dataset using `max()` and `summarize()`:

Reviewer: ```suggestion
> > Identify the most recent year in our dataset using `max()` and `summarize()`:
```

> >
> > ```{r FilterChallenge1}
> > smoking %>%
> > summarize(recent_year=max(year))
> > ```
> >
> > Filter to only the maximum year, and determine the average percent of people with lung cancer.
> >
> > ```{r FilterChallenge2}
> > smoking %>%
> > filter(year == 2019) %>%
> > summarize(mean_cancer_pct=mean(lung_cancer_pct))
> > ```
> > {: .source}
> > By combining `filter()` and `summarize()` we were able to calculate the mean mercent of people with lung cancer in 2019
> > 
> > Bonus:

> > > Instead of finding the minimum year first and then filtering, we can find the minimum year right in the filter function: 

Reviewer: ```suggestion
> > Instead of finding the most recent year first and then filtering, we can find the most recent year right in the filter function: 
```

> > ```{r no_hard_code}
> > smoking %>% 

> > > filter(year == min(year)) %>% 

Reviewer: ```suggestion
> > filter(year == max(year)) %>% 
```

> > summarize(mean_cancer_pct=mean(lung_cancer_pct))
> > ```
> {: .solution}
{: .challenge}

## Grouping rows using `group_by()` {#grouping-rows-using-group_by}
[*Back to top*](#contents)

> We see that the mean percent of smokers is larger in 1990 than the value we got using all of the rows. But now we might be interested in calculating the average for each year. Rather that doing a bunch of different `filter()` statements, we can instead use the `group_by()` function. The function allows us to tell the code to treat the rows in logical groups, so rather than summarizing over all the rows, we will get one summary value for each group. Here's what that will look like:

Reviewer: ```suggestion
We previously saw that the mean percent of smokers is larger in 1990 than the value we got using all of the rows. But now we might be interested in calculating the average for each year. Rather that doing a bunch of different `filter()` statements, we can instead use the `group_by()` function. The function allows us to treat the rows in logical groups, so rather than summarizing over all the rows, we will get one summary value for each group. Here's what that will look like:
```

```{r GapLifeExpByYear}
smoking %>%
  group_by(year) %>%
  summarize(average=mean(smoke_pct))
```

The `group_by()` function expects you to pass in the name of a column (or multiple columns separated by commas) in your data. 

Note that you might get a message about the summarize function regrouping the output by 'year'. This simply indicates what the function is grouping by. 

> > ## Pro tip: Counting

Reviewer: Do you want to also add count() or tally() as alternatives to summarize here?

> 
> Often we want to count how many of each category we have in our dataset. 
> To do this, we can use the function `n()` within `summarize()`. 
> For instance, here we count the number of countries in each continent for which we have data in 2019:
> 
> ```{r}
> smoking %>% 
>   filter(year == 2019) %>% 
>   group_by(continent) %>% 
>   summarize(count = n())
> ```

> > 

Author: ```suggestion
> 
> Other alternatives are the `count()` and `tally()` functions, which are shortcuts. For instance:
> smoking %>% 
>   filter(year == 2019) %>% 
>   group_by(continent) %>% 
>   count()
```

{: .callout}

> ## Grouping the data
>
> Try calculating the average percent of smokers by continent.
>
> > ## Solution
> >
> > ```{r GroupByChallenge1}
> > smoking %>%
> > group_by(continent) %>%
> > summarize(avg_smoke_pct=mean(smoke_pct))
> > ```
> > {: .source}
> >
> > By combining `group_by()` and `summarize()` we are able to calculate the mean percent of smokers by continent.
> {: .solution}
{: .challenge}

> ## Bonus: plotting review
>
> Make a line plot of year vs. average percent of smokers to be able to more easily visualize the relationship between these two variables. 
>
> > ## Solution
> >
> > ```{r}
> > smoking %>%
> >   group_by(year) %>%
> >   summarize(average=mean(smoke_pct)) %>% 
> >   ggplot(aes(x = year, y = average)) +
> >   geom_line()
> > ```
> > {: .source}
> >
> > You can pipe your summarized data right to ggplot! 
> {: .solution}
{: .challenge}

> 

Reviewer: Okay, I know we're doing a lot in here, but I feel like these are some of THE most important skills to teach. What do you think of an example where we group by two variables? Maybe year and continent? That would be manageable output, I think?

## Make new variables with `mutate()` {#make-new-variables-with-mutate}
[*Back to top*](#contents)

Each time we ran `summarize()`, we got back fewer rows than passed in. We either got one row back, or one row per group. But sometimes we want to create a new column in our data without changing the number of rows. The function we use to create new columns is called `mutate()`.

We have a column for the population and the percent of people who smoke. If we wanted to get an estimate of the number of people who smoke, we could multiply the percentages by the total population. Here's what such a `mutate()` command would look like:

```{r GapMutate}
smoking %>%

>   mutate(smoke_n = pop * smoke_pct)

Reviewer: Pretty sure we changed these to actual percentages, right? (aka multiplied by 100)
```suggestion
  mutate(smoke_n = pop * smoke_pct/100)
```

```

This will add a new column called "smoke_n" to our data. We use the column names as if they were regular values that we want to perform mathematical operations on and provide the name in front of an equals sign like we have done with `summarize()`

> ## `mutate()`
> We can also multiply by constants or other numbers using mutate - remember how in the plotting lesson we made a plot with population in millions? Try making a new column for this dataframe called pop_mil that is the population in millions. 
> 
> > ## Solution: 
> > ```{r mutateExercise}  
> > smoking %>%  

> > > mutate(smoke_n = pop * smoke_pct, pop_mil = pop / 1000000)  

Reviewer: ```suggestion
> > mutate(smoke_n = pop * smoke_pct/100, pop_mil = pop / 1000000)  
```

> The `select()` function has a bunch of helper functions that are handy if you are working with a dataset that has a lot of columns. You can see these helper functions on the `?select` help page. For example, let's say we wanted to select the year column and all the columns that start with the letter "c". You can do that with:
> 
> ```{r GapSelectFancy}
> smoking %>%
>   select(year, starts_with("c"))
> ```

> > This returns just the three columns we are interested in. 

Reviewer: ```suggestion
> This returns just the three columns we are interested in. Note that we had to use quotation marks for "c" because the `starts_with()` function requires a character vector in the match argument.
```

The `tidyr` package contains the functions `pivot_wider()` and `pivot_longer()` that make it easy to switch between the two formats. The `tidyr` package is included in the `tidyverse` package so we don't need to do anything to load it.

First, our data is not as clean as it looks - there are some countries that have multiple entries for a given year. 
However, we want only one entry per year. 
To get this, we will summarize the data to be the mean of all values for a given year:

```{r PivotWider}
smoking %>%
  group_by(country, continent, year) %>% 
  summarize(smoke_pct = mean(smoke_pct))
```

Notice how we went from having `r nrow(smoking)` rows to `r smoking %>% group_by(country, continent, year) %>% summarize(smoke_pct = mean(smoke_pct)) %>% nrow()` in our data. 

Next, let's create a wide version of the data using `pivot_wider()`:

```{r}
smoking %>%
  group_by(country, continent, year) %>% 
  summarize(smoke_pct = mean(smoke_pct)) %>%

>   select(country, continent, year, smoke_pct) %>%

Reviewer: ```suggestion
```

  pivot_wider(names_from = year, values_from = smoke_pct)
```

Notice here that we tell `pivot_wider()` which columns to pull the names we wish our new columns to be named from the year variable, and the values to populate those columns from the `smoke_pct` variable. (Again, neither of which have to be in quotes in the code when there are no special characters or spaces - certainly an incentive not to use special characters or spaces!) We see that the resulting table has new columns by year, and the values populate it with our remaining variables dictating the rows.

## Plotting wide data

Let's make a plot with our wide data comparing percent of smokers in 1990 to percent of smokers in 2010 to see how it has changed for each country.

```{r}
smoking %>%
  group_by(country, continent, year) %>% 
  summarize(smoke_pct = mean(smoke_pct)) %>%

>   select(country, continent, year, smoke_pct) %>%

Reviewer: ```suggestion
```

  pivot_wider(names_from = year, values_from = smoke_pct) %>% 
  ggplot(aes(x = 1990, y = 2010)) +
  geom_point()
```

> Hmm that's not what we want. `ggplot` just plotted the numbers 1952 and 2007 instead of the data from the years. That's because it evaluates those as numbers instead of column names. To fix this, we can add a prefix to the years in `pivot_wider()`:

Reviewer: ```suggestion
Hmm that's not what we want. `ggplot` just plotted the numbers 1990 and 2010 instead of the data from the years. That's because it evaluates those as numbers instead of column names. To fix this, we can add a prefix to the years in `pivot_wider()`:
```

```{r}
smoking %>%
  group_by(country, continent, year) %>% 
  summarize(smoke_pct = mean(smoke_pct)) %>%

>   select(country, continent, year, smoke_pct) %>%

Reviewer: ```suggestion
```

  pivot_wider(names_from = year, values_from = smoke_pct, names_prefix = 'y') %>% 
  ggplot(aes(x = y1990, y = y2010)) +
  geom_point()
```

> That's better, but the different ranges on the x and y axis make it hard to compare the points. 

Reviewer: ```suggestion
Alright, now we have a plot with the mean percent of smokers in in 1990 on the x axis and the mean percent of smokers in 2010 on the y axis, and each point represents a country. However, the different ranges on the x and y axis make it hard to compare the points. 
```

Let's fix that by adding a line at y=x. 

```{r}
smoking %>%
  group_by(country, continent, year) %>% 
  summarize(smoke_pct = mean(smoke_pct)) %>%

>   select(country, continent, year, smoke_pct) %>%

Reviewer: ```suggestion
```

  pivot_wider(names_from = year, values_from = smoke_pct, names_prefix = 'y') %>% 
  ggplot(aes(x = y1990, y = y2010)) +
  geom_point() +
  geom_abline(intercept = 0, slope = 1)
```

> It seems like in most countries the percent of smokers has decrased from 1990 to 2010, but there are some countries where smoking has increased. Let's figure out which those are!

Reviewer: ```suggestion
It seems like in most countries the percent of smokers has decreased from 1990 to 2010, since most of the points fall below the line y = x. However, there are some countries where smoking has increased (i.e. the points are above the line y = x). Let's figure out which those are!
```

> ## Bonus: Identifying countries with more smokers in 2010 than 1990
>
> Use what you've learned from today to figure out which countries had higher smoking percentage in 2010 than 1990. 
>
> Bonus: Order the data frame from greatest to smallest difference. HINT: The `arrange()` function can help you do this. 
> 
> > ## Solution
> > ```{r}
> > smoking %>%

> > >   group_by(country, continent, year) %>% # select the columns you want to keep

Reviewer: ```suggestion
> >   group_by(country, continent, year) %>% # group by the columns you want to keep
```
